src/firebaseConfig.ts

```diff
@@ -5,74 +5,146 @@
 // 'npm run generate:json-schema' to regenerate the schema files.
 //
 
-export type DatabaseConfig =
+// Sourced from - https://firebase.google.com/docs/hosting/full-config#rewrite-cloud-run-container
+type CloudRunRegions =
+  | "asia-east1"
+  | "asia-east2"
+  | "asia-northeast1"
+  | "asia-northeast2"
+  | "asia-northeast3"
+  | "asia-south1"
+  | "asia-southeast1"
+  | "asia-southeast2"
+  | "australia-southeast1"
+  | "europe-north1"
+  | "europe-west1"
+  | "europe-west2"
+  | "europe-west3"
+  | "europe-west4"
+  | "europe-west6"
+  | "northamerica-northeast1"
+  | "southamerica-east1"
+  | "us-central1"
+  | "us-east1"
+  | "us-east4"
+  | "us-west1";
+
+type CloudFunctionRuntimes = "nodejs10" | "nodejs12" | "nodejs14";
+
+type DeployAsset = {
+  predeploy?: string | string[];
+  postdeploy?: string | string[];
+};
+
+type DatabaseOne = {
+  rules: string;
+} & DeployAsset;
+
+type DatabaseMany = ({
+  rules: string;
+} & (
   | {
-      rules?: string;
+      instance: string;
+      target?: string;
     }
   | {
-      target?: string;
       instance?: string;
-      rules: string;
-    }[];
+      target: string;
+    }
+) &
+  DeployAsset)[];
 
-export type FirestoreConfig = {
-  rules?: string;
-  indexes?: string;
+type HostingRedirects = ({ source: string } | { regex: string }) & {
+  destination: string;
+  type: number;
 };
 
-export type FunctionsConfig = {
-  // TODO: Add types for "backend" and "runtime"
-  source?: string;
-  ignore?: string[];
-  predeploy?: string[];
+type HostingRewrites = ({ source: string } | { regex: string }) &
+  (
+    | { destination: string }
+    | { function: string }
+    | {
+        run: {
+          serviceId: string;
+          region?: CloudRunRegions;
+        };
+      }
+    | { dynamicLinks: boolean }
+  );
+
+type HostingHeaders = ({ source: string } | { regex: string }) & {
+  headers: {
+    key: string;
+    value: string;
+  }[];
 };
 
-export type HostingConfig = {
+type HostingBase = {
   public: string;
   ignore?: string[];
   appAssociation?: string;
   cleanUrls?: boolean;
   trailingSlash?: boolean;
-  postdeploy?: string;
-  redirects?: {
-    source: string;
-    destination: string;
-    type: number;
-  }[];
-  rewrites?: {
-    source: string;
-    destination?: string;
-    function?: string;
-    run?: {
-      serviceId: string;
-      region?: string;
-    };
-    dynamicLinks?: boolean;
-  }[];
-  headers?: {
-    source: string;
-    headers: {
-      key: string;
-      value: string;
-    }[];
-  }[];
+  redirects?: HostingRedirects[];
+  rewrites?: HostingRewrites[];
+  headers?: HostingHeaders[];
   i18n?: {
     root: string;
   };
 };
 
-export type StorageConfig =
-  | {
-      rules: string;
-    }
-  | {
-      bucket: string;
-      rules: string;
-    }[];
+type HostingOne = HostingBase &
+  DeployAsset & {
+    site?: string;
+    target?: string;
+  };
+
+type HostingMany = (HostingBase &
+  DeployAsset &
+  (
+    | {
+        site: string;
```

Comment:
Seems like in a lot of places you're using a `RequireAtLeastOne` type to make sure at least one field of a set is defined. Maybe we can use some TS wizardry for this?

https://docs.microsoft.com/en-us/javascript/api/@azure/keyvault-certificates/requireatleastone?view=azure-node-latest

It's a pretty nasty type definition but we can just define it once up top, link to the docs, and then simplify a lot of the other types.

Reply:
I used this type. It is only actually used in two places.

Reply:
Thanks!